client/ayon_houdini/plugins/publish/extract_last_published.py

```diff
@@ -0,0 +1,107 @@
+
+from ayon_core.lib import collect_frames
+from ayon_houdini.api import plugin
+
+
+class ExtractLastPublished(plugin.HoudiniExtractorPlugin):
+    """
+    Extractor that copies files from last published
+    to staging directory.
+    It works only if instance data includes "last_version_published_files"
+    and there are frames to fix.
+
+    The files from last published are base of files which will be extended/fixed for specific
+    frames.
+
+    NOTE: 
+        This plugin is closely taken from ayon-nuke.
+        It contains some Houdini addon specific logic as various addons may
+          have unique methods for managing `staging_dir`, `expectedFiles`, and `frames`.
+    TODO:
+        It's preferable to to generalize this plugin for broader use and
+          integrate it into ayon-core.
+    """
+
+    order = pyblish.api.ExtractorOrder - 0.1
+    label = "Extract Last Published"
+    targets = ["local"]  # Same target as `CollectFramesFixDef`
+    families = ["*"]
+
+    def process(self, instance):
+        frames_to_fix = instance.data.get("frames_to_fix")
+        last_published = instance.data.get("last_version_published_files")
+        if not last_published:
+            self.log.debug("Skipping, No last publish found.")
+            return 
+        if not frames_to_fix :
+            self.log.debug("Skipping, No frames to fix.")
+            return
+        last_published_and_frames = collect_frames(last_published)
+        
+        if not all(last_published_and_frames.values()):
+            self.log.debug("Skipping, No file sequence found in the "
+                           "last version published files.")
+            return
+
+        staging_dir, expected_filenames = self.get_expected_files_and_staging_dir(instance)
+
+        os.makedirs(staging_dir, exist_ok=True)
+
+        expected_and_frames = collect_frames(expected_filenames)
+        frames_and_expected = {v: k for k, v in expected_and_frames.items()}
+        frames_to_fix = clique.parse(frames_to_fix, "{ranges}")
+        
+        anatomy = instance.context.data["anatomy"]
+
+        for file_path, frame in last_published_and_frames.items():
+            if frame is None:
+                continue
+            file_path = anatomy.fill_root(file_path)
+            if not os.path.exists(file_path):
+                continue
+            target_file_name = frames_and_expected.get(frame)
+            if not target_file_name:
+                continue
+            
+            out_path = os.path.join(staging_dir, target_file_name)
+
+            # Copy only the frames that we won't render.
+            if frame and frame not in frames_to_fix:
+                self.log.debug("Copying '{}' -> '{}'".format(file_path, out_path))
+                shutil.copy(file_path, out_path)
+
+
+    def get_expected_files_and_staging_dir(self, instance):
+        """ Get expected file names or frames.
+
+        This method includes Houdini specific code.
+        """
```

Comment:
Can we get arguments and return with types in the docstring?

Reply:
Thank you. Here it's.
https://github.com/ynput/ayon-houdini/pull/45/commits/8cef57a3c46eb3277b1dd5d9542aea3587eb055b